bin/opd8n/src/cmd/from_l2.rs

```diff
@@ -18,6 +48,231 @@ pub struct FromL2 {
+                    }
+                },
+            }
+
+            // Get the attributes if there are some available.
+            let Some(attributes) = pipeline.next() else {
+                continue;
+            };
+
+            // Print the L1 range for this L2 Block.
+            let derived = attributes.parent.block_info.number as i64 + 1;
+            let l2_block_info = l2_provider
+                .l2_block_info_by_number(derived as u64)
+                .await
+                .map_err(|e| eyre!(e))?;
+            let origin = pipeline
+                .origin()
+                .ok_or(eyre!("Failed to get pipeline l1 origin"))?;
+            info!(target: TARGET,
+                "L2 Block [{}] L1 Range: [{}, {}]",
+                derived, l2_block_info.l1_origin.number, origin.number
+            );
+            payloads.push(attributes.attributes);
+            last_l1_block = origin.number;
+
+            // Keep trying to advance the cursor in case the fetch fails.
+            loop {
+                match l2_provider
+                    .l2_block_info_by_number(l2_cursor.block_info.number + 1)
+                    .await
+                {
+                    Ok(bi) => {
+                        l2_cursor = bi;
+                        break;
+                    }
+                    Err(e) => {
+                        error!(target: TARGET, "Failed to fetch next pending l2 safe head: {}, err: {:?}", l2_cursor.block_info.number + 1, e);
+                        // Don't step on the pipeline if we failed to fetch the next l2 safe head.
+                        continue;
+                    }
+                }
+            }
+        }
+
+        // Take the full L1 range of blocks and get all needed data.
+        let l1_blocks = (first_l1_block..=last_l1_block).collect::<Vec<u64>>();
+
+        // Construct the fixture blocks.
+        let blocks = crate::cmd::build_fixture_blocks(
+            cfg.batch_inbox_address,
+            cfg.genesis
+                .system_config
+                .as_ref()
+                .map(|sc| sc.batcher_address)
+                .unwrap_or_default(),
+            &l1_blocks,
+            &mut l1_provider,
+            &mut blob_provider,
+        )
+        .await?;
+        let fixture = DerivationFixture::new(blocks, payloads);
+        info!(target: TARGET, "Successfully built derivation test fixture");
+
+        // Write the derivation fixture to the specified output location.
+        let file = std::fs::File::create(&self.output)?;
+        serde_json::to_writer_pretty(file, &fixture)?;
+        info!(target: "from-l1", "Wrote derivation fixture to: {:?}", self.output);
+
+        Ok(())
+    }
+
+    /// Gets the L2 starting block number.
+    /// Returns the genesis L2 block number if the start block is less than the genesis block number.
+    pub fn start_block(&self, cfg: &RollupConfig) -> u64 {
```

Comment:
Is there a way to combine `FromL1` and `FromL2` into a single `FromBlocks` struct? 

It looks like both `FromL1` and `FromL2` have the same cli args as well as some duplicate code. Maybe we could pass an additonal arg that specifies `--from l1` or `--from l2`?